File: packages/kotlin/src/main/kotlin/dev/naturalspacing/core/NaturalSpacingSession.kt
package dev.naturalspacing.core

class NaturalSpacingSession {
    private var suppressions = mutableListOf<SuppressedBoundary>()
    private var lastPolicy: FieldPolicy? = null

    val suppressedBoundaryCount: Int get() = suppressions.size

    fun reset() {
        suppressions.clear()
        lastPolicy = null
    }

    fun process(snapshot: EditSnapshot): EditPlan {
        if (lastPolicy != null && lastPolicy != snapshot.policy) suppressions.clear()
        lastPolicy = snapshot.policy
        rebase(snapshot)
        if (snapshot.policy == FieldPolicy.NATURAL_LANGUAGE && snapshot.composingRange == null) {
            deletedBoundary(snapshot)?.let { boundary ->
                suppressions.removeAll { it.offset == boundary.offset }
                suppressions += boundary
            }
        }
        val planned = NaturalSpacing.planEdit(snapshot, suppressions.map { it.offset }.toSet())
        val plan = if (planned.decision == PlanDecision.NO_CHANGE && suppressions.isNotEmpty()) {
            planned.copy(decision = PlanDecision.SUPPRESSED)
        } else planned
        suppressions = suppressions.mapNotNull { suppression ->
            val mapped = NaturalSpacing.mapSelection(
                TextSelection(suppression.offset, suppression.offset),
                plan.insertions,
            ).anchor
            boundaryContext(plan.resultText, mapped)?.takeIf {
                it.left == suppression.left && it.right == suppression.right
            }?.let { suppression.copy(offset = mapped) }
        }.toMutableList()
        return plan
    }

    private fun rebase(snapshot: EditSnapshot) {
        val start = snapshot.changedRange.start
        val end = start + snapshot.changedRange.length
        val delta = snapshot.afterUserText.length - snapshot.beforeText.length
        suppressions = suppressions.mapNotNull { suppression ->
            var offset = suppression.offset
            if (offset > end || (snapshot.changedRange.length > 0 && offset == end)) offset += delta
            else if (offset > start && offset < end) return@mapNotNull null
            boundaryContext(snapshot.afterUserText, offset)?.takeIf {
                it.left == suppression.left && it.right == suppression.right
            }?.let { suppression.copy(offset = offset) }
        }.toMutableList()
    }
}

private data class SuppressedBoundary(val offset: Int, val left: String, val right: String)
private data class BoundaryContext(val left: String, val right: String, val leftCategory: Category, val rightCategory: Category)

private fun deletedBoundary(snapshot: EditSnapshot): SuppressedBoundary? {
    if (snapshot.editKind != EditKind.DELETE || snapshot.changedRange.length != 1 ||
        snapshot.beforeText.length - 1 != snapshot.afterUserText.length ||
        snapshot.beforeText.substring(snapshot.changedRange.start, snapshot.changedRange.start + 1) != " "
    ) return null
    val context = boundaryContext(snapshot.afterUserText, snapshot.changedRange.start) ?: return null
    if (NaturalSpacing.insertionReason(context.leftCategory, context.rightCategory) == null) return null
    return SuppressedBoundary(snapshot.changedRange.start, context.left, context.right)
}

private fun boundaryContext(text: String, offset: Int): BoundaryContext? {
    val graphemes = NaturalSpacing.segment(text)
    val rightIndex = graphemes.indexOfFirst { it.start == offset }
    if (rightIndex <= 0) return null
    val left = graphemes[rightIndex - 1]
    val right = graphemes[rightIndex]
    if (left.end != offset) return null
    return BoundaryContext(left.text, right.text, left.category, right.category)
}

File: packages/dotnet/NaturalSpacing.Core/NaturalSpacingSession.cs
namespace NaturalSpacing.Core;

public sealed class NaturalSpacingSession
{
    private List<SuppressedBoundary> _suppressions = [];
    private FieldPolicy? _lastPolicy;

    public int SuppressedBoundaryCount => _suppressions.Count;

    public void Reset()
    {
        _suppressions.Clear();
        _lastPolicy = null;
    }

    public EditPlan Process(EditSnapshot snapshot)
    {
        if (_lastPolicy is not null && _lastPolicy != snapshot.Policy) _suppressions.Clear();
        _lastPolicy = snapshot.Policy;
        Rebase(snapshot);
        if (snapshot.Policy == FieldPolicy.NaturalLanguage && snapshot.ComposingRange is null &&
            DeletedBoundary(snapshot) is { } boundary)
        {
            _suppressions.RemoveAll(item => item.Offset == boundary.Offset);
            _suppressions.Add(boundary);
        }

        var planned = NaturalSpacingFormatter.PlanEdit(snapshot, _suppressions.Select(item => item.Offset).ToHashSet());
        var plan = planned.Decision == PlanDecision.NoChange && _suppressions.Count > 0
            ? planned with { Decision = PlanDecision.Suppressed }
            : planned;
        _suppressions = _suppressions.Select(suppression =>
        {
            var mapped = NaturalSpacingFormatter.MapSelection(
                new TextSelection(suppression.Offset, suppression.Offset), plan.Insertions).Anchor;
            var context = BoundaryContext(plan.ResultText, mapped);
            return context is not null && context.Left == suppression.Left && context.Right == suppression.Right
                ? suppression with { Offset = mapped }
                : null;
        }).Where(item => item is not null).Select(item => item!).ToList();
        return plan;
    }

    public ProposedEditResult ProcessProposedEdit(ProposedEdit edit) =>
        NaturalSpacingFormatter.ProcessProposedEdit(edit, Process);

    private void Rebase(EditSnapshot snapshot)
    {
        var start = snapshot.ChangedRange.Start;
        var end = start + snapshot.ChangedRange.Length;
        var delta = snapshot.AfterUserText.Length - snapshot.BeforeText.Length;
        _suppressions = _suppressions.Select(suppression =>
        {
            var offset = suppression.Offset;
            if (offset > end || (snapshot.ChangedRange.Length > 0 && offset == end)) offset += delta;
            else if (offset > start && offset < end) return null;
            var context = BoundaryContext(snapshot.AfterUserText, offset);
            return context is not null && context.Left == suppression.Left && context.Right == suppression.Right
                ? suppression with { Offset = offset }
                : null;
        }).Where(item => item is not null).Select(item => item!).ToList();
    }

    private static SuppressedBoundary? DeletedBoundary(EditSnapshot snapshot)
    {
        if (snapshot.EditKind != EditKind.Delete || snapshot.ChangedRange.Length != 1 ||
            snapshot.BeforeText.Length - 1 != snapshot.AfterUserText.Length ||
            snapshot.BeforeText.Substring(snapshot.ChangedRange.Start, 1) != " ") return null;
        var context = BoundaryContext(snapshot.AfterUserText, snapshot.ChangedRange.Start);
        return context is not null && NaturalSpacingFormatter.Reason(context.LeftCategory, context.RightCategory) is not null
            ? new SuppressedBoundary(snapshot.ChangedRange.Start, context.Left, context.Right)
            : null;
    }

    private static Boundary? BoundaryContext(string text, int offset)
    {
        var graphemes = NaturalSpacingFormatter.Segment(text);
        var rightIndex = -1;
        for (var index = 0; index < graphemes.Count; index++)
            if (graphemes[index].Start == offset) { rightIndex = index; break; }
        if (rightIndex <= 0) return null;
        var left = graphemes[rightIndex - 1];
        var right = graphemes[rightIndex];
        return left.End == offset ? new Boundary(left.Text, right.Text, left.Category, right.Category) : null;
    }

    private sealed record SuppressedBoundary(int Offset, string Left, string Right);
    private sealed record Boundary(string Left, string Right, Category LeftCategory, Category RightCategory);
}
